plugin/pkg/admission/initialresources/admission.go

```diff
@@ -0,0 +1,154 @@
+		Handler: admission.NewHandler(admission.Create),
+		source:  source,
+	}
+}
+
+func (ir initialResources) Admit(a admission.Attributes) (err error) {
+	// Ignore all calls to subresources or resources other than pods.
+	if a.GetSubresource() != "" || a.GetResource() != string(api.ResourcePods) {
+		return nil
+	}
+	pod, ok := a.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	ir.estimateResourcesIfNotSet(pod)
+	return nil
+}
+
+func (ir initialResources) estimateResourcesIfNotSet(pod *api.Pod) {
+	for i := range pod.Spec.Containers {
+		c := &pod.Spec.Containers[i]
+		req := c.Resources.Requests
+		lim := c.Resources.Limits
+		var cpu, mem *resource.Quantity
+		var err error
+		if _, ok := req[api.ResourceCPU]; !ok {
+			if _, ok2 := lim[api.ResourceCPU]; !ok2 {
+				cpu, err = ir.getEstimation(api.ResourceCPU, c)
+				if err != nil {
+					glog.Errorf("Error while trying to estimate resources: %v", err)
+				}
+			}
+		}
+		if _, ok := req[api.ResourceMemory]; !ok {
+			if _, ok2 := lim[api.ResourceMemory]; !ok2 {
+				mem, err = ir.getEstimation(api.ResourceMemory, c)
+				if err != nil {
+					glog.Errorf("Error while trying to estimate resources: %v", err)
+				}
+			}
+		}
+
+		// If Requests doesn't exits and an estimation was made, create Requests.
+		if req == nil && (cpu != nil || mem != nil) {
+			c.Resources.Requests = api.ResourceList{}
+			req = c.Resources.Requests
+		}
+		if cpu != nil {
+			glog.Infof("CPU estimation for container %v in pod %v/%v is %v", c.Name, pod.ObjectMeta.Namespace, pod.ObjectMeta.Name, cpu.String())
+			req[api.ResourceCPU] = *cpu
+		}
+		if mem != nil {
+			glog.Infof("Memory estimation for container %v in pod  %v/%v is %v", c.Name, pod.ObjectMeta.Namespace, pod.ObjectMeta.Name, mem.String())
+			req[api.ResourceMemory] = *mem
+		}
+	}
+	// TODO(piosz): verify the estimates fits in LimitRanger
+}
+
+func (ir initialResources) getEstimation(kind api.ResourceName, c *api.Container) (*resource.Quantity, error) {
+	end := time.Now()
+	start := end.Add(-week)
+	var usage, samples int64
+	var err error
+
+	// Historical data from last 7 days for the same image:tag.
+	if usage, samples, err = ir.source.GetUsagePercentile(kind, *percentile, c.Image, true, start, end); err != nil {
+		return nil, err
+	}
+	if samples < 60 {
+		// Historical data from last 30 days for the same image:tag.
+		start := end.Add(-month)
+		if usage, samples, err = ir.source.GetUsagePercentile(kind, *percentile, c.Image, true, start, end); err != nil {
+			return nil, err
+		}
+	}
+	if samples < 60 {
+		// Historical data from last 30 days for the same image.
+		start := end.Add(-month)
+		image := strings.Split(c.Image, ":")[0]
+		if usage, samples, err = ir.source.GetUsagePercentile(kind, *percentile, image, false, start, end); err != nil {
+			return nil, err
+		}
+	}
+
+	if samples > 0 && kind == api.ResourceCPU {
```

Comment:
This is how it is described in proposal doc. If you think there should be a different logic here I'm happy to discuss but let's make it in a proposal and then apply changes here.